csrc/scheduler/matmul_utils.cpp

```diff
@@ -0,0 +1,794 @@
+  // TODO: add proper set of checks
+  return "";
+}
+
+std::string getMatmulCompileTimeRejectReason(Fusion* fusion) {
+  // The plan:
+  // 1. check if there is exactly one MmaOp defined in the fusion
+  // 2. check if MmaOp inputs match any of supported layout
+  // 3. check if fusion represents expressions that are recognized by matmul
+  //    scheduler
+
+  // #1
+  const auto fusion_exprs = fusion->exprs();
+  auto mma_exprs = ir_utils::filterByType<MmaOp>(fusion_exprs).vector();
+  if (mma_exprs.size() != 1) {
+    std::stringstream ss;
+    ss << "Matmul scheduler supports only a single MMA op in the fusion, got: "
+       << mma_exprs.size();
+    return ss.str();
+  }
+
+  // #2
+  {
+    for (const auto* mma_expr : mma_exprs) {
+      const auto layout_data = getLayout(mma_expr);
+      if (layout_data.second) {
+        return layout_data.second.value();
+      }
+    }
+  }
+
+  // #3
+  {
+    for (auto mma_expr : mma_exprs) {
+      const auto matmul_type_v2 = checkMatmulType(fusion, mma_expr);
+      if (!matmul_type_v2.second->empty()) {
+        return matmul_type_v2.second.value();
+      }
+    }
+  }
+
+  return "";
+}
+
+std::shared_ptr<MatmulParams> getMatmulHeuristics(
+    Fusion* fusion,
+    SchedulerRuntimeInfo& runtime_info,
+    HeuristicSummary* data_cache) {
+  FusionGuard fg(fusion);
+  (void)data_cache;
+  (void)runtime_info;
+  auto params = std::make_shared<MatmulParams>();
+
+  // Check initial conditions
+  const auto fusion_exprs = fusion->exprs();
+  auto mma_exprs = ir_utils::filterByType<MmaOp>(fusion_exprs).vector();
+  if (mma_exprs.size() != 1) {
+    // Support only for fusion with a single mma op
+    return nullptr;
+  }
+
+  const auto layout = getLayout(mma_exprs.front());
+  if (layout.second) {
+    // Layout check returned an error message
+    if (isDebugDumpEnabled(DebugDumpOption::MatmulChecks)) {
+      printMsg(layout.second.value());
+    }
+    return nullptr;
+  }
+
+  const auto problem_shape = getProblemShape(
+      fusion, mma_exprs[0]->as<MmaOp>(), runtime_info, layout.first.value());
+  if (!problem_shape) {
+    // Failed to acquire problem shape
+    return nullptr;
+  }
+
+  const auto device_prop = at::cuda::getCurrentDeviceProperties();
+  const auto mma_op = getMmaOp(
+      device_prop->major * 10 + device_prop->minor, problem_shape.value());
+  if (!mma_op) {
+    // No heuristics can be prepared if mma op request is empty
+    return nullptr;
+  }
+
+  // Populate heuristic details
+  auto status = initCoreHeuristics(
+      params, mma_op.value(), layout.first.value(), problem_shape.value());
+  if (!status) {
+    // Core part of heuristics failed to initialize
+    return nullptr;
+  }
+
+  status = initExtraHeuristics(params, problem_shape.value());
+  if (!status) {
+    // Additional pieces of heuristics failed to initialize
+    return nullptr;
+  }
+
+  // set kernel index mode
+  params->cparams.index_type = getIndexType(problem_shape.value());
```

Comment:
Hmmm, all the other schedulers are using  `indexModeToDtype(runtime_info.getIndexMode());`, if matmul is having trouble with it, then very likely all other schedulers are also having trouble with it. I am OK with your `getIndexType` approach today, but I think I should investigate this issue deeper.

Reply:
Yes, I was right. Other schedulers are also affected: https://github.com/NVIDIA/Fuser/issues/79